### PR TITLE
feat(ai): tool-driven LLM with github-state MCP server (#117)

### LINKED ISSUE #117: feat(ai): unified tool-driven LLM architecture — github-state MCP server + tools across reasoning executors
## Summary

Move all reasoning LLM calls in this codebase to a unified **tool-driven** architecture, with a new `github-state` MCP server. Today the bot's LLM calls receive a fixed pre-fetched context and cannot fetch fresh GitHub state on demand — leading to vague, hallucinated answers.

## Motivating Incident

On https://github.com/chrisleekr/personal-claw/pull/25 a user asked `@chrisleekr-bot-dev why this is not automerging?`. The chat-thread executor replied with a generic five-bullet checklist (`"CI checks haven't passed (or are still running)"`) instead of the definitive answer: `"the Lint, Type Check & Test and auto-merge jobs are FAILURE on this PR — see <link>"`. PR #25 actually had 2 FAILURE check runs the bot could not see.

Root cause: `src/workflows/ship/scoped/chat-thread.ts:418-431` ships only `state`, `draft`, `base_ref`, `head_ref` to the LLM. The in-source comment admits the gap:

> `// PR state block — TRUSTED. v1: minimal. The full PR-state-aware probe (CI status, unresolved threads, behind-base) is a follow-up`

## Current Inventory

| Executor / classifier | Today | Reasons about state? | Action |
|---|---|---|---|
| `chat-thread` (`src/workflows/ship/scoped/chat-thread.ts`) | single-turn, no tools | yes | migrate to tool-driven |
| `fix-thread` (`src/workflows/ship/scoped/fix-thread.ts`) | Agent SDK + Bash + MCP, no CI awareness | yes | register `github-state` |
| `investigate` (`src/workflows/ship/scoped/investigate.ts`) | Agent SDK | yes | register `github-state` |
| `summarize` (`src/workflows/ship/scoped/summarize.ts`) | Agent SDK | yes | register `github-state` |
| `open-pr` (`src/workflows/ship/scoped/open-pr.ts`) | Agent SDK | yes | register `github-state` |
| legacy `pipeline.ts` (`src/core/pipeline.ts`) | Agent SDK with prompt-stuffed PR data | yes | register `github-state`, drop pre-fetch |
| `triage` (`src/orchestrator/triage.ts`) | single-turn workload classifier | yes — accuracy improves with PR size/CI | migrate to tool-driven |
| `nl-classifier` (`src/workflows/ship/nl-classifier.ts`) | single-turn intent classifier | no — text-only | leave as-is |
| `intent-classifier` (`src/workflows/intent-classifier.ts`) | single-turn intent classifier | no — text-only | leave as-is |
| `probe` / `verdict` (`src/workflows/ship/probe.ts`, `verdict.ts`) | **deterministic** GraphQL → boolean | n/a | leave as-is — correctness invariant |

Note: `runProbe` already fetches `statusCheckRollup` deterministically for the merge loop. The new MCP tools should reuse that GraphQL where possible to avoid duplicating the query.

### 1. `src/mcp/servers/github-state.ts` — read-only MCP server

Initial tool surface:

- `get_pr_state_check_rollup(owner, repo, number)` — rollup state + ranked rows + `is_required`
- `get_check_run_output(owner, repo, check_run_id)` — `output.summary` + `output.text` (truncated) + `html_url`
- `get_workflow_run(owner, repo, run_id)` — conclusion + `logs_url` + `html_url`
- `get_branch_protection(owner, repo, branch)` — required checks list
- `get_pr_diff(owner, repo, number)` — capped patch text
- `list_pr_comments(owner, repo, number, cursor?)` — paginated

All responses token-bounded + paginated. Auth via the existing daemon `installationToken` (same path as `inline-comment` MCP). Follows existing MCP registry pattern.

### 2. `src/ai/llm-client.ts` — `runWithTools()`

New entry point supporting Anthropic's `tools` + `tool_choice: "auto"`. Runs the tool loop server-side, capped at **4 tool-call iterations per turn** (configurable, fail-open if exceeded). Final response still passes through `parseStructuredResponse(raw, schema)` for executors that expect structured JSON.

Single-turn LLM consumers (`chat-thread`, `triage`) migrate to this. Multi-turn Agent SDK consumers register the MCP server and let the SDK handle the loop natively.

### 3. Carve-outs

- `nl-classifier`, `intent-classifier` — text classification, no GitHub fetch needed; keep single-turn no-tools.
- `probe`, `verdict` — deterministic; LLM nondeterminism would regress merge-readiness correctness.

## Why This Is More Robust Than Prompt-Stuffing

- Same data path for single-turn and multi-turn executors → one tool surface to test, one auth path.
- LLM decides what to fetch — engineer doesn't pre-decide every field every executor needs.
- Extending to a new GitHub surface (deployments, dependabot alerts) = add one tool, every executor inherits.
- No staleness — every fetch is fresh.
- Token cost lower in steady state because data is fetched only when relevant.

## Phased Rollout

| Phase | Scope | Approx LOC |
|---|---|---|
| 0 | `runWithTools` in `llm-client.ts` (no consumers yet) | ~120 |
| 1 | `github-state` MCP server (read-only, 4 starter tools) | ~250 |
| 2 | Wire `chat-thread` → `runWithTools` + register `github-state` (fixes PR #25 scenario) | ~80 |
| 3 | Wire `triage` → `runWithTools` + register `github-state` | ~50 |
| 4 | Wire Agent SDK executors (`fix-thread`, `investigate`, `summarize`, `open-pr`, `pipeline`) | ~100 |
| 5 | Extend tool surface (branch-protection, diff, comments, deployments) | additive |

Each phase ships as an independent PR.

## Risks

1. **Tool-call thrash** — model may over-call. Mitigation: 4/turn cap, tight tool descriptions, observability counter per call site.
2. **Latency** — CI-relevant chat turn becomes ~+1.5–3s (one extra round trip). Acceptable for a maintainer bot.
3. **Cost** — ~1.5–2× per CI-relevant turn. Multi-tenant Console path needs per-installation budget guard.
4. **Test surface** — build `MockGithubStateServer` once, share across executor tests.
5. **`probe` regression vector** — never let the LLM decide mergeability; correctness lives in the deterministic projection.

## Open Questions

- Should the MCP server live in-process (server-side) or out-of-process (subprocess)? Existing MCP servers in this repo follow the subprocess pattern.
- Does `runWithTools` share Bedrock + OAuth + API-key auth paths the same way `runStructured` does today?
- Per-installation tool-call budget — env-var threshold or DB row?

## Acceptance Criteria

- [ ] `@chrisleekr-bot-dev why isn't this merging?` on a PR with failing CI returns a definitive answer naming the failing checks and linking to logs.
- [ ] `triage` accuracy improves on PRs where the comment alone is ambiguous but the PR state is informative (file count, CI state).
- [ ] No regression in `probe` / merge-readiness verdicts.
- [ ] Per-turn tool-call cap enforced and observable.
- [ ] Test mock available for executor tests.

## Related

- `src/workflows/ship/scoped/chat-thread.ts:418-431` — the source of the gap
- `src/workflows/ship/probe.ts`, `src/workflows/ship/verdict.ts` — existing deterministic CI-aware fetcher to reuse GraphQL from
- `src/ai/llm-client.ts` — single entry point that today routes Anthropic / OAuth / Bedrock; new `runWithTools` lands here
- `src/ai/structured-output.ts` — `parseStructuredResponse` chokepoint stays load-bearing for tool-loop final output

### PR DESCRIPTION
**Pull request by chrisleekr**
## Summary

Closes the gap that surfaced on `chrisleekr/personal-claw#25` — `@chrisleekr-bot-dev why this is not automerging?` returned a generic five-bullet checklist instead of the definitive answer because `chat-thread` shipped only `state`, `draft`, `base_ref`, `head_ref` to the LLM and could not fetch CI rollup on demand.

This PR introduces a unified tool-driven path. A new `runWithTools()` loop drives Anthropic's tool-use protocol for single-turn callers (`chat-thread`, orchestrator `triage`); a new read-only `github-state` MCP subprocess surfaces the same tool surface to Agent SDK callers (`pipeline.ts`). Both dispatch paths share one fetcher implementation in `src/github/state-fetchers.ts`, so adding a tool means editing one file and every consumer inherits.

The static `<pr_state>` block in chat-thread is removed; a `<freshness_directive>` block in `prompt-builder.ts` tells the agent to prefer tools for current state and use the snapshot only for stable metadata.

## Diagram

Before:

```mermaid
flowchart LR
  ChatThread["chat-thread.ts"] -->|"static pr_state:<br/>state, draft, base_ref, head_ref"| LLMCreate["llm.create<br/>single turn"]
  Pipeline["pipeline.ts"] -->|"prompt-stuffed<br/>PR snapshot"| AgentSDK["Agent SDK<br/>no fresh-state path"]
  Triage["orchestrator/triage.ts"] -->|"owner, repo,<br/>labels, body"| LLMCreate2["llm.create"]
  classDef gap fill:#7a3434,color:#ffffff
  class LLMCreate,LLMCreate2,AgentSDK gap
```

After:

```mermaid
flowchart LR
  ChatThread["chat-thread.ts<br/>config.chatThreadToolsEnabled"] -->|"GITHUB_STATE_TOOLS<br/>onToolCall"| RunWithTools["runWithTools<br/>iterations cap 8<br/>fan-out cap 4<br/>fail-open"]
  Triage["triage.ts<br/>config.triageToolsEnabled"] -->|"iterations cap 2"| RunWithTools
  RunWithTools --> CreateRich["LLMClient.createRich<br/>tool_use blocks"]
  CreateRich --> Dispatch["dispatchGithubStateTool"]
  Pipeline["pipeline.ts<br/>enableGithubState"] -->|"register MCP<br/>subprocess"| McpServer["github-state<br/>MCP subprocess<br/>thin wrapper"]
  Dispatch --> Fetchers["state-fetchers.ts<br/>get_pr_state_check_rollup<br/>get_check_run_output<br/>get_workflow_run<br/>get_branch_protection<br/>get_pr_diff<br/>get_pr_files<br/>list_pr_comments"]
  McpServer --> Fetchers
  classDef good fill:#1e6f3f,color:#ffffff
  class RunWithTools,Fetchers,McpServer good
```

## Changes

**`src/ai/llm-client.ts` — `runWithTools` loop and `createRich` API**

- New `LLMClient.createRich(params)` returns content blocks + `stop_reason` (tool-use-aware sibling of `create`).
- New `runWithTools(client, params)` drives the tool loop server-side: dispatches every `tool_use` block, appends `tool_result` blocks, loops until `end_turn` or cap.
- `DEFAULT_MAX_TOOL_ITERATIONS = 8` (per-turn round-trip cap); `DEFAULT_MAX_TOOL_USES_PER_TURN = 4` (per-turn fan-out cap). Excess `tool_use` blocks get `is_error: true` `tool_result` feedback so the model can adjust on the next turn.
- Tool handler errors are caught per-call and surfaced as `is_error` tool results; loop continues so the model can recover.
- Cap-exceeded path fail-opens with the latest assistant text plus `capExceeded: true` and `droppedToolCalls` counters.
- OAuth array-form system blocks (Claude Code identifier first) extracted to shared `buildSystemField` so `buildRequest` and `buildRichRequest` stay in lockstep.

**`src/github/queries.ts` (new)**

- Extracted `PROBE_QUERY` and `REVIEW_THREADS_PAGE_QUERY` from `src/workflows/ship/probe.ts`. One source of truth between the deterministic merge-readiness probe and the new MCP tool surface.

**`src/github/state-fetchers.ts` (new) — shared tool implementations**

- Seven read-only fetchers: `getPrStateCheckRollup`, `getCheckRunOutput`, `getWorkflowRun`, `getBranchProtection`, `getPrDiff`, `getPrFiles`, `listPrComments`.
- `GITHUB_STATE_TOOLS` descriptors mirror the MCP server's advertised surface 1:1.
- `dispatchGithubStateTool(deps, call)` switch is the single-turn dispatcher passed as `onToolCall`.
- Per-tool truncation (60KB text, 50KB diff, 30KB check output) bounds the conversation budget.
- Rollup rows sorted with failed+required first so a model that truncates can still answer the common question.

**`src/mcp/servers/github-state.ts` (new) — Agent SDK tool surface**

- Thin stdio wrapper around the same fetchers (mirrors existing `inline-comment.ts` pattern).
- Hard-pinned to current repo via `REPO_OWNER` + `REPO_NAME` env so the model cannot fan out to arbitrary repos.
- Registered in `src/mcp/registry.ts` via `enableGithubState` opt-in flag; bundle entrypoint added to `scripts/build.ts`.

**`src/workflows/ship/scoped/chat-thread.ts` — wired to runWithTools**

- Static `<pr_state>` block dropped from `buildUserPrompt`.
- PR conversations get `tools: GITHUB_STATE_TOOLS` + `onToolCall: dispatchGithubStateTool`. Issue conversations stay tool-less (PR-scoped tools).
- `CHAT_THREAD_SYSTEM_PROMPT` extended with tool descriptions and usage rules pinned on the motivating incident.
- Gated by new `config.chatThreadToolsEnabled` (env: `CHAT_THREAD_TOOLS_ENABLED`, default `true`).

**`src/orchestrator/triage.ts` — tool-aware triage**

- New optional `octokit` + `prNumber` fields on `TriageInput` for tool dispatch on PR events.
- Dedicated `TRIAGE_MAX_TOOL_ITERATIONS = 2` (lower than chat-thread's 8) bounds hot-path latency.
- `SYSTEM_PROMPT` updated with "call AT MOST ONE tool per classification" guidance.
- Gated by new `config.triageToolsEnabled` (env: `TRIAGE_TOOLS_ENABLED`, default `true`).
- `webhook/router.ts` passes `octokit` + `entityNumber` when `ctx.isPR`.

**`src/core/pipeline.ts` — Agent SDK registration**

- New `overrides.enableGithubState` (defaults `true`); registers MCP server + appends seven `mcp__github_state__*` tool names to `allowedTools`.

**`src/core/prompt-builder.ts` — `<freshness_directive>`**

- Added between `<security_directive>` and `<formatted_context>`. Tells the agent the snapshot is point-in-time; prefer `mcp__github_state__*` tools for current state; use the snapshot for stable metadata and review comments (no tool covers those yet).

**`src/workflows/dispatcher.ts` and `src/workflows/ship/scoped/dispatch-scoped.ts` — tools-aware callLlm adapters**

- Both `buildCallLlm` builders branch on `tools` + `onToolCall` and dispatch via `runWithTools`, fall back to `client.create` otherwise. Backward compatible with all other single-turn callers (`investigate`, `summarize`, `nl-classifier`, etc.).

**Tests**

- `test/ai/run-with-tools.test.ts` — 16 tests covering loop semantics: `end_turn` short-circuit, single-tool round trip, fail-open on cap exceeded, `DEFAULT_MAX_TOOL_USES_PER_TURN = 4`, fan-out cap dispatching + `is_error` feedback for excess, handler error → `is_error` tool_result, multi-tool single-turn dispatch, usage summing, request shaping (tools + tool_choice forms, OAuth array-form system blocks), and `parseAnthropicRichResponse` edge cases (interleaved blocks, unknown stop_reason, malformed tool_use drop).
- `test/github/state-fetchers.test.ts` — 9 tests covering tool surface declaration (7 tools, schema-shaped inputs), parameter validation rejection paths, sorted rollup with failed+required first, `protected: false` on 404, `get_pr_files` happy path, non-404 errors → `is_error`.

**Carve-outs honoured**

- `src/workflows/ship/probe.ts` and `verdict.ts` — deterministic merge-readiness path, untouched (correctness invariant).
- `src/workflows/intent-classifier.ts` and `src/workflows/ship/nl-classifier.ts` — text-only classifiers, unchanged.

**Docs**

- `docs/build/extending.md` — new section on sharing tool surfaces between MCP subprocess and single-turn `runWithTools` callers.
- `scripts/build.ts` — new `github-state.ts` entrypoint so `dist/mcp/servers/github-state.js` ships in production.

## Related Issues

- Closes #117

## Test plan

- [x] Tested locally — 25/25 new tests pass; 263 pre-existing failures unchanged.
- [x] Added/updated tests — 16 new for `runWithTools`, 9 for `dispatchGithubStateTool`.
- [x] All existing tests pass — net zero regressions vs `main` baseline (confirmed via `git stash` diff: 264 fail on main, 263–265 with PR depending on test ordering noise; failures are in unrelated suites: `intent-classifier`, `k8s`, `dispatch-stats`, `fetcher`).
- [x] `bun run typecheck` clean.
- [x] `bun run lint` 0 errors (343 pre-existing warnings unchanged).
- [x] `bun run audit:ci` 0 blocking.
- [x] `bun run build` emits `dist/mcp/servers/github-state.js`.
- [x] `bun run docs:build` clean (mkdocs --strict).
- [x] `scripts/check-docs-citations.ts` and `scripts/check-docs-versions.ts` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/github/state-fetchers.ts [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+
+function truncate(text: string, max: number): { text: string; truncated: boolean } {
+  if (text.length <= max) return { text, truncated: false };
+  return { text: text.slice(0, max), truncated: true };
+}
+
+function serialize(value: unknown, max = MAX_TEXT_BYTES): string {
+  const raw = JSON.stringify(value);
+  const { text, truncated } = truncate(raw, max);
+  return truncated ? `${text} /* …truncated */` : text;
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/github/state-fetchers.ts]: Fixed in 2951f1f: serialize() now returns a valid-JSON envelope ({truncated:true, original_byte_length, cap_bytes}) instead of slicing the string and appending a non-JSON suffix.

File: src/github/state-fetchers.ts [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+      branch,
+      protected: true,
+      required_status_checks: result.data.required_status_checks ?? null,
+      required_pull_request_reviews: result.data.required_pull_request_reviews ?? null,
+      enforce_admins: result.data.enforce_admins?.enabled ?? false,
+      restrictions: result.data.restrictions ?? null,
+    });
+  } catch (err) {
+    // 404 is expected for unprotected branches — return a structured payload, not an error.
+    if (err instanceof Error && /404|not.found/i.test(err.message)) {
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/github/state-fetchers.ts]: Fixed in 2951f1f: switched to err.status === 404 (Octokit RequestError shape) instead of message-string regex.

File: src/github/state-fetchers.ts [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+          return { content: JSON.stringify({ error: "pr_number required" }), isError: true };
+        }
+        return { content: await getPrFiles(deps, input.pr_number) };
+      }
+      case "list_pr_comments": {
+        const input = call.input as { pr_number?: unknown; page?: unknown };
+        if (typeof input.pr_number !== "number") {
+          return { content: JSON.stringify({ error: "pr_number required" }), isError: true };
+        }
+        const page = typeof input.page === "number" ? input.page : 1;
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/github/state-fetchers.ts]: Fixed in 2951f1f: page is now validated as a positive integer; non-integer / 0 / negative inputs return an is_error tool_result before any GitHub call.

File: src/orchestrator/triage.ts [hunk 4]
```diff
@@ -230,17 +264,43 @@ export async function triageRequest(input: TriageInput, client: LLMClient): Prom
...
+        runWithTools(client, {
+          model: modelId,
+          system: withStructuredRules(SYSTEM_PROMPT),
+          messages: [{ role: "user", content: prompt }],
+          maxTokens: config.triageMaxTokens,
+          tools: GITHUB_STATE_TOOLS,
+          maxIterations: TRIAGE_MAX_TOOL_ITERATIONS,
+          onToolCall: (call) =>
+            dispatchGithubStateTool({ octokit, owner: input.owner, repo: input.repo }, call),
+        }),
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/orchestrator/triage.ts:289]: Fixed in 2951f1f: buildTriagePrompt now emits a PR number line when input.prNumber is supplied so the model can pass pr_number to the github-state tools.

File: src/core/pipeline.ts [hunk 5]
```diff
@@ -270,15 +285,28 @@ export async function runPipeline(
...
+            ...withResolveTool,
+            "mcp__github_state__get_pr_state_check_rollup",
+            "mcp__github_state__get_check_run_output",
+            "mcp__github_state__get_workflow_run",
+            "mcp__github_state__get_branch_protection",
+            "mcp__github_state__get_pr_diff",
+            "mcp__github_state__get_pr_files",
+            "mcp__github_state__list_pr_comments",
+          ]
+        : withResolveTool;
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/core/pipeline.ts]: Fixed in 2951f1f: enableGithubState now defaults to ctx.isPR; issue-only workflows no longer spawn the MCP server or expose PR-shaped tools that would just error out.

File: src/mcp/servers/github-state.ts [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+async function runServer(): Promise<void> {
+  const transport = new StdioServerTransport();
+  await server.connect(transport);
+  process.on("exit", () => {
+    void server.close();
+  });
+}
+
+void runServer().catch((err: unknown) => {
+  console.error(err);
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/mcp/servers/github-state.ts:176]: Fixed in 2951f1f: runServer().catch now process.exit(1) on connect failure (mirrors resolve-review-thread.ts) so a dead sidecar surfaces to the supervisor.
